### PR TITLE
[SHOTS] Allows a list of shots to be specified

### PR DESCRIPTION
**Pull request by josh146**
**Context:** Expands the functionality of the QNode, by allowing measuring statistics to be course grained over shots in a single evaluation.

**Description of the changes:**

Consider

```pycon
>>> shots_list = [5, 10, 1000]
>>> dev = qml.device("default.qubit", wires=2, analytic=False, shots=shots_list)
```

When QNodes are executed on this device, a single execution of 1015 shots will be submitted.
However, three sets of measurement statistics will be returned; using the first 5 shots,
second set of 10 shots, and final 1000 shots, separately.

For example:

```python
@qml.qnode(dev)
def circuit(x):
    qml.RX(x, wires=0)
    qml.CNOT(wires=[0, 1])
    return qml.expval(qml.PauliZ(0) @ qml.PauliX(1)), qml.expval(qml.PauliZ(0))
```

Executing this, we will get an output of size `(3, 2)`:

```pycon
>>> circuit(0.5)
[[0.33333333 1.        ]
 [0.2        1.        ]
 [0.012      0.868     ]]
```

The output remains fully differentiable.

**Benefits:**

* More fine-grained control over the shots budget, support for shot-adaptive optimizers
* The logic is relatively well-contained, so it should be easy to update this to follow any UI changes to how shots are specified.

**Drawbacks:** Breaks all plugins that depend on `QubitDevice`, due to two new keyword arguments in `expval`, `var`, and `sample`.

### CODE REVIEW

**Comment by mariaschuld** [comment]
Nice! 

One thought, we cannot assume that other devices inheriting from  `Device` support sequences of shots. What will happen if this is attempted in those devices?

File: pennylane/_device.py [hunk 1]
```diff
@@ -178,13 +178,45 @@ def shots(self, shots):
...
+                    # in a sparse format.
+                    shot_vector = [(shots[0], len(shots))]
+                    total_shots = shots[0] * len(shots)
+                else:
+                    # Iterate through the shots, and group consecutive identical shots
+                    split_at_repeated = np.split(shots, np.where(np.diff(shots) != 0)[0] + 1)
+                    shot_vector = [
+                        i.item() if len(i) == 1 else (i[0], len(i)) for i in split_at_repeated
+                    ]
+                    total_shots = np.sum(shots)
```
Comment (mariaschuld) [inline pennylane/_device.py:206]: Could outsource this procedure to a function, so it can be easily tested?

File: pennylane/_qubit_device.py [hunk 2]
```diff
@@ -543,13 +578,15 @@ def analytic_probability(self, wires=None):
...
-    def estimate_probability(self, wires=None):
+    def estimate_probability(self, wires=None, shot_range=None, bin_size=None):
         """Return the estimated probability of each computational basis state
         using the generated samples.
 
         Args:
             wires (Iterable[Number, str], Number, str, Wires): wires to calculate
                 marginal probabilities for. Wires not provided are traced out of the system.
+            shot_range (tuple[int]): 2-tuple of integers specifying the range of samples
+                to use. If not specified, all samples are used.
```
Comment (mariaschuld) [inline pennylane/_qubit_device.py:589]: Also add bin_size arg ?

File: pennylane/_device.py (hunk 1 above)
Comment (josh146) [reply pennylane/_device.py:206]: Good idea, will add!

File: pennylane/_qubit_device.py (hunk 2 above)
Comment (josh146) [reply pennylane/_qubit_device.py:589]: Oops, forgot a few docstring arguments I can see.

I'm still a bit wary of adding new kwargs, but it feels safer than attributes.

**Comment by co9olguy** [comment]
Is `analytic=False` still needed in the example?

**Comment by josh146** [comment]
> Is analytic=False still needed in the example?

@co9olguy yes, until #1079 is merged, which removes the `analytic` keyword.

File: pennylane/_qubit_device.py [hunk 3]
```diff
@@ -561,19 +601,33 @@ def estimate_probability(self, wires=None):
...
+        if bin_size is not None:
+            bins = len(samples) // bin_size
+
+            indices = indices.reshape((bins, -1))
+            prob = np.zeros([2 ** len(device_wires), bins], dtype=np.float64)
+
+            # count the basis state occurrences, and construct the probability vector
+            for b, idx in enumerate(indices):
+                basis_states, counts = np.unique(idx, return_counts=True)
+                prob[basis_states, b] = counts / bin_size
```
Comment (josh146) [inline pennylane/_qubit_device.py:621]: I wonder if this can be vectorized, rather than a for loop? Perhaps it's also possible to do away with the if/else, and have a single block of logic for any bin size.

**Comment by josh146** [comment]
> One thought, we cannot assume that other devices inheriting from Device support sequences of shots. What will happen if this is attempted in those devices?

For devices that don't know about `self._shot_vector`, nothing changes - they will continue to just use the total number of shots. Don't know if this makes sense or not - we could add something to the capabilities dictionary?

File: pennylane/_device.py [hunk 4]
```diff
@@ -36,6 +36,48 @@
...
+    if all(isinstance(s, int) for s in shot_list):
+
+        if len(set(shot_list)) == 1:
+            # All shots are identical; represent the shot vector
+            # in a sparse format.
+            shot_vector = [(shot_list[0], len(shot_list))]
+            total_shots = shot_list[0] * len(shot_list)
+        else:
+            # Iterate through the shots, and group consecutive identical shots
+            split_at_repeated = np.split(shot_list, np.where(np.diff(shot_list) != 0)[0] + 1)
```
Comment (Thenerdstation) [inline pennylane/_device.py:69]: I'm having a hard time parsing this line.

So the first `np.diff(shot_list) != 0` is a boolean array with the first indicies where the shot value changes marked as True, and `np.where(...)` returns tuple of an array of the indices of these first values. Is that correct?

Comment (Thenerdstation) [inline pennylane/_device.py:69]:
```suggestion
            split_at_repeated = np.split(shot_list, np.diff(shot_list).nonzero()[0] + 1)
```
As per the Note here:
https://numpy.org/doc/stable/reference/generated/numpy.where.html?highlight=where#numpy.where

File: pennylane/_device.py [hunk 5]
```diff
@@ -36,6 +36,48 @@
...
+    sequence is returned, where tuples indicate the number of times a shot
+    integer is repeated.
+    """
+    if all(isinstance(s, int) for s in shot_list):
+
+        if len(set(shot_list)) == 1:
+            # All shots are identical; represent the shot vector
+            # in a sparse format.
+            shot_vector = [(shot_list[0], len(shot_list))]
+            total_shots = shot_list[0] * len(shot_list)
```
Comment (Thenerdstation) [inline pennylane/_device.py:66]: This chunk isn't needed as the code below should correctly handle this case

File: pennylane/_device.py [hunk 6]
```diff
@@ -36,6 +36,48 @@
...
+            # Iterate through the shots, and group consecutive identical shots
+            split_at_repeated = np.split(shot_list, np.where(np.diff(shot_list) != 0)[0] + 1)
+            shot_vector = [i.item() if len(i) == 1 else (i[0], len(i)) for i in split_at_repeated]
+            total_shots = np.sum(shot_list)
+
+    else:
+        # shot_list is already a shot vector, simply compute the total number of shots
+        shot_vector = shot_list
+        total_shots = sum(i if isinstance(i, int) else np.prod(i) for i in shot_list)
+
```
Comment (Thenerdstation) [inline pennylane/_device.py:77]: I feel like we should instead error instead in this case, especially since we aren't doing any additional checks on shots_list

**Review by Thenerdstation — COMMENTED** [review]
I'm not feeling super keen on this right now. It feels like we're adding a ton of complexity for what amounts to a for loop in user code.

File: pennylane/_qubit_device.py [hunk 7]
```diff
@@ -208,7 +208,34 @@ def execute(self, circuit, **kwargs):
...
+
+                if isinstance(shot, tuple):
+                    s2 = s1 + np.prod(shot)
+                    bin_size = shot[0]
+                else:
+                    s2 = s1 + shot
+                    bin_size = shot
+
+                r = self.statistics(circuit.observables, shot_range=[s1, s2], bin_size=bin_size)
+                r = np.squeeze(r)
```
[the remark this answers is not in the document]
Comment (josh146) [reply pennylane/_qubit_device.py:226]: I didn't here because backprop is 'broken' once you sample, but I think you are right, we would have to to ensure the JIT isn't broken

File: pennylane/_qubit_device.py [hunk 8]
```diff
@@ -208,7 +208,34 @@ def execute(self, circuit, **kwargs):
...
+            s1 = 0
+
+            for shot in self._shot_vector:
+
+                if isinstance(shot, tuple):
+                    s2 = s1 + np.prod(shot)
+                    bin_size = shot[0]
+                else:
+                    s2 = s1 + shot
+                    bin_size = shot
```
[the remark this answers is not in the document]
Comment (josh146) [reply pennylane/_qubit_device.py:223]: True!

File: pennylane/_device.py (hunk 5 above)
Comment (josh146) [reply pennylane/_device.py:66]: True, I added this after for two reasons:

* I think `shots=[n] * N` is probably the most common use-case
* In the case of `shots=[n] * N`, this is faster than the second branch.

But yes, the second branch should handle both cases

**Comment by josh146** [comment]
> I'm not feeling super keen on this right now. It feels like we're adding a ton of complexity for what amounts to a for loop in user code.

Unfortunately, a for loop in the user code (while giving the same results) will generate separate jobs that may be held up in a remote queue. The idea here is to 'batch' shots on a single QNode eval.

Shot adaptive quantum optimizers typically need to generate many 1-shot expectation values, which is currently not possible to implement in PennyLane without submitting many single-shot jobs

File: pennylane/_device.py [hunk 9]
```diff
@@ -36,6 +36,48 @@
...
+    Returns:
+        tuple[int, list[int, tuple[int]]]: A tuple containing the total number
+        of shots, as well as the shot vector.
+
+    **Example**
+
+    >>> shot_list = [3, 1, 2, 2, 2, 2, 6, 1, 1, 5, 12, 10, 10]
+    >>> _process_shot_sequence(shot_list)
+    (57, [3, 1, (2, 4), 6, (1, 2), 5, 12, (10, 2)])
```
[the remark this answers is not in the document]
Comment (josh146) [reply pennylane/_device.py:54]: Makes sense!

File: pennylane/_device.py (hunk 6 above)
Comment (josh146) [reply pennylane/_device.py:77]: Have added a validation here!

File: pennylane/_qubit_device.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (josh146) [reply pennylane/_qubit_device.py:226]: Have replaced this with `qml.math.squeeze`

File: pennylane/_qubit_device.py [hunk 10]
```diff
@@ -208,7 +208,34 @@ def execute(self, circuit, **kwargs):
...
+                r = np.squeeze(r)
+
+                if isinstance(shot, tuple):
+                    results.extend(r.T)
+                else:
+                    results.append(np.array(r))
+
+                s1 = s2
+
+            results = np.stack(results)
```
[the remark this answers is not in the document]
Comment (josh146) [reply pennylane/_qubit_device.py:235]: Have replaced this with `qml.math.stack`

**Comment by josh146** [comment]
@Thenerdstation have standardized the format to a namedtuple, and made every element a tuple - it cleans it up a _lot_, and helps remove a lot of redundant lines.

File: pennylane/_qubit_device.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (josh146) [reply pennylane/_qubit_device.py]: Removed this if statement by having the type be consistent

File: pennylane/_device.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (josh146) [reply pennylane/_device.py]: Done both suggestions.

File: pennylane/_device.py (hunk 4 above)
Comment (josh146) [reply pennylane/_device.py]: Yep, essentially. This line splits the shots list, grouping together consecutive and indentical elements:

image

File: pennylane/_device.py [hunk 11]
```diff
@@ -36,6 +36,48 @@
 from pennylane.wires import Wires, WireError
 
 
+def _process_shot_sequence(shot_list):
+    """Process the shot sequence, to determine the total
+    number of shots and the shot vector.
+
+    Args:
+        shot_list (Sequence[int, tuple[int]]): sequence of non-negative shot integers
```
[the remark this answers is not in the document]
Comment (josh146) [reply pennylane/_device.py:48]: At the back of my mind, I wanted to allow (advanced) users and other parts of the codebase to pass `shots=[(1, 1e9)]`, rather than having to do `shots=[1] * 1000000000`, which will create a large list in memory for absolutely no reason

File: pennylane/_qubit_device.py (hunk 3 above)
Comment (josh146) [reply pennylane/_qubit_device.py:617]: 😢

**Review by Thenerdstation — APPROVED** [review]
This feels a lot cleaner thank you!

Approval for code structure and readability, but I leave it to nathan and maria to decide if we want to break the existing plugins for this.

**Comment by josh146** [comment]
> This feels a lot cleaner thank you!

Np!

> Approval for code structure and readability, but I leave it to nathan and maria to decide if we want to break the existing plugins for this.

I expect #1079 to pre-break all plugins, so took a little liberty here 😆

**Comment by Thenerdstation** [comment]
> I expect #1079 to pre-break all plugins, so took a little liberty here

That's fair

File: pennylane/_device.py [hunk 12]
```diff
@@ -178,13 +178,45 @@ def shots(self, shots):
...
+                    "The specified number of shots needs to be at least 1. Got {}.".format(shots)
+                )
+
+            self._shots = int(shots)
+            self._shot_vector = None
+
+        elif isinstance(shots, Sequence):
+            # a sequence of shots has been provided
+
+            if all(isinstance(s, int) for s in shots):
```
Comment (mariaschuld) [inline pennylane/_device.py]: I wonder if we could drop this check, since it iterates through the entire list? I know that this will only happen once at device creation, but if users provide wrong inputs, the error should be thrown where it breaks (as long as one can legibly  figure out what was wrong), no?

File: pennylane/_device.py [hunk 13]
```diff
@@ -36,6 +36,62 @@
...
+
+    The total number of shots (57), and a sparse representation of the shot
+    sequence is returned, where tuples indicate the number of times a shot
+    integer is repeated.
+    """
+    if all(isinstance(s, int) for s in shot_list):
+
+        if len(set(shot_list)) == 1:
+            # All shots are identical, only require a single shot tuple
+            shot_vector = [ShotTuple(shots=shot_list[0], copies=len(shot_list))]
```
Comment (mariaschuld) [inline pennylane/_device.py:76]: Interesting, I didn't know named tuples!

File: pennylane/_device.py [hunk 14]
```diff
@@ -36,6 +36,62 @@
...
+        else:
+            # Iterate through the shots, and group consecutive identical shots
+            split_at_repeated = np.split(shot_list, np.diff(shot_list).nonzero()[0] + 1)
+            shot_vector = [ShotTuple(shots=i[0], copies=len(i)) for i in split_at_repeated]
+
+    elif all(isinstance(s, (int, tuple)) for s in shot_list):
```
Comment (mariaschuld) [inline pennylane/_device.py:82]: the int part of the check would have been picked up by the branch above, no? so is it superfluous here?

File: pennylane/_qubit_device.py [hunk 15]
```diff
@@ -71,8 +71,9 @@ class QubitDevice(Device):
         wires (int, Iterable[Number, str]]): Number of subsystems represented by the device,
             or iterable that contains unique labels for the subsystems as numbers (i.e., ``[-1, 0, 2]``)
             or strings (``['ancilla', 'q1', 'q2']``). Default 1 if not specified.
-        shots (int): number of circuit evaluations/random samples used to estimate
-            expectation values of observables
+        shots (int, list[int]): Number of circuit evaluations/random samples used to estimate
+            expectation values of observables. If a list of integers is passed, the circuit
+            evaluations are batched over the list of shots.
```
Comment (mariaschuld) [inline pennylane/_qubit_device.py:76]: I think this will need some more explanation? For example, what is the rule for circuit outputs? If the device is analytic and `qml.probs(), qml.sample()` is returned, will only the second return value be batched? But will both be batched if the device is non-analytic?

I know that this will somewhat change by shots refactor II, but good to be precise?

Comment (josh146) [reply pennylane/_qubit_device.py:76]: Yes, good point. At the moment, anything that relies on samples will get batched if the shots are batched. So that includes samples always, and expval/var/probs in non-analytic mode.

This will definitely be more intuitive after II, because the presence of batched shots will _always_ correspond to batched output.

File: pennylane/_device.py (hunk 14 above)
Comment (josh146) [reply pennylane/_device.py:82]: I don't think so, since that check was `all int` and this check is `all (int or tuple)`. So the first could still fail while the second passes.

File: pennylane/_device.py (hunk 12 above)
Comment (josh146) [reply pennylane/_device.py]: I wasn't sure how much validation to do here 🤔 

For `shots=n`, we check not only if `n` is an integer, but if `n>1`.

File: pennylane/_qubit_device.py [hunk 16]
```diff
@@ -543,13 +574,18 @@ def analytic_probability(self, wires=None):
...
         using the generated samples.
 
         Args:
             wires (Iterable[Number, str], Number, str, Wires): wires to calculate
                 marginal probabilities for. Wires not provided are traced out of the system.
+            shot_range (tuple[int]): 2-tuple of integers specifying the range of samples
+                to use. If not specified, all samples are used.
+            bin_size (int): Divides the shot range into bins of size ``bin_size``, and
+                returns the measurement statistic separately over each bin. If not
+                provided, the entire shot range is treated as a single bin.
```
Comment (mariaschuld) [inline pennylane/_qubit_device.py:588]: I am slow on the uptake here...what is the role of shot_range vs bin_size and how do they relate to the shots_vector?

File: tests/test_qubit_device.py [hunk 17]
```diff
@@ -861,3 +861,124 @@ def test_result_empty_tape(self, mock_qubit_device_with_paulis_and_methods, tol)
...
+    def test_multiple_probs(self, shot_list, shot_vector, expected_shape, total_shots):
+        """Test multiple probability returns"""
+        dev = qml.device("default.qubit", wires=2, analytic=False, shots=shot_list)
+
+        @qml.qnode(dev)
+        def circuit(x, y):
+            qml.RX(x, wires=0)
+            qml.RY(y, wires=0)
+            qml.CNOT(wires=[0, 1])
+            return qml.probs(wires=0), qml.probs(wires=1)
```
Comment (mariaschuld) [inline tests/test_qubit_device.py:967]: I am actually not sure now, but do we allow for mixed-measurement-process outputs like `probs(), expval()`? If so, isn't that an important edge case?

**Review by mariaschuld — COMMENTED** [review]
Nice one @Josh! 

Just a few questions before approving, but it looks good to me.

File: tests/test_qubit_device.py [hunk 18]
```diff
@@ -861,3 +861,124 @@ def test_result_empty_tape(self, mock_qubit_device_with_paulis_and_methods, tol)
...
+        # test gradient works
+        res = qml.jacobian(circuit)(0.5, 0.1)
+
+    def test_invalid_shot_list(self):
+        """Test exception raised if the shot list is the wrong type"""
+        with pytest.raises(qml.DeviceError, match="Shots must be"):
+            qml.device("default.qubit", wires=2, analytic=False, shots=0.5)
+
+        with pytest.raises(ValueError, match="Unknown shot sequence"):
+            qml.device("default.qubit", wires=2, analytic=False, shots=["a", "b", "c"])
```
Comment (mariaschuld) [inline tests/test_qubit_device.py:984]: Do we need a test for the per-qnode-call way of setting shots?

File: pennylane/_qubit_device.py (hunk 16 above)
Comment (josh146) [reply pennylane/_qubit_device.py:588]: Ah, so say you have `shots=[5, (10, 3), 100]`.

For the second element (10 shots, 3 copies) we will want to use samples number 5 to 35, performing the expectation value in bins of size 10.

So in this case `shot_range=[5, 35]`, `bin_size=10`.

Maybe there are better argument names I could use?

Comment (mariaschuld) [reply pennylane/_qubit_device.py:588]: Aah! no, that makes sense.

File: tests/test_qubit_device.py (hunk 17 above)
Comment (josh146) [reply tests/test_qubit_device.py:967]: This is undefined behaviour at the moment 🤔 This results in a ragged array, and I purposely didn't try to solve it due to an upcoming change in NumPy that will make it really hard to work with object arrays
